module/input/PS3Controller/src/PS3Controller.cpp

```diff
@@ -0,0 +1,244 @@
+        ACCELEROMETER_X = 23,
+        ACCELEROMETER_Y,
+        ACCELEROMETER_Z
+    };
+
+    using extension::Configuration;
+
+    using message::input::PS3Controller::Accelerometer;
+    using message::input::PS3Controller::CircleButton;
+    using message::input::PS3Controller::CrossButton;
+    using message::input::PS3Controller::DPadButton;
+    using message::input::PS3Controller::L1Button;
+    using message::input::PS3Controller::L2Trigger;
+    using message::input::PS3Controller::LeftJoystick;
+    using message::input::PS3Controller::LeftJoystickButton;
+    using message::input::PS3Controller::PSButton;
+    using message::input::PS3Controller::R1Button;
+    using message::input::PS3Controller::R2Trigger;
+    using message::input::PS3Controller::RightJoystick;
+    using message::input::PS3Controller::RightJoystickButton;
+    using message::input::PS3Controller::SelectButton;
+    using message::input::PS3Controller::SquareButton;
+    using message::input::PS3Controller::StartButton;
+    using message::input::PS3Controller::TriangleButton;
+
+    PS3Controller::PS3Controller(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+
+        on<Configuration>("PS3Controller.yaml").then([this](const Configuration& config) {
+            // Only reconnect if we are changing the path to the device
+            if ((controller_path.compare(config["controller_path"]) != 0)
+                || (accelerometer_path.compare(config["accelerometer_path"]) != 0)) {
+                controller_path    = config["controller_path"];
+                accelerometer_path = config["accelerometer_path"];
+                connect();
+            }
+        });
+
+        on<Shutdown>().then([this] { disconnect(); });
+
+        // Details on Linux Joystick API
+        // https://www.kernel.org/doc/Documentation/input/joystick-api.txt
+        // Trigger when the joystick has an data to read
+        on<IO>(controller_fd, IO::READ).then([this] {
+            // JoystickEvent has
+            // Timestamp: uint32_t
+            // Value:     int16_t
+            // Type:      uint8_t
+            // Number:    uint8_t
+            constexpr size_t JOYSTICK_EVENT_SIZE =
+                sizeof(uint32_t) + sizeof(int16_t) + sizeof(uint8_t) + sizeof(uint8_t);
+
+            // Try to read a JoystickEvent worth of data
+            std::array<uint8_t, JOYSTICK_EVENT_SIZE> buffer;
+            auto num_bytes = read(controller_fd, buffer.data(), buffer.size());
+            if (num_bytes > -1) {
+                for (const uint8_t& byte : buffer) {
+                    event_buffer.push_back(byte);
+                }
+            }
+
+            while (event_buffer.size() >= JOYSTICK_EVENT_SIZE) {
+                // Extract timestamp from event buffer in milliseconds and convert it to a NUClear::clock::time_point
+                auto timestamp = std::chrono::time_point<NUClear::clock, std::chrono::milliseconds>(
```

Comment:
Are we sure this is the right clock for this? It might be steady clock vs system clock etc. it might be that a duration makes more sense

Reply:
you tell me, all I know is its a timestamp (`event timestamp in milliseconds`)

Reply:
Well then it would be wrong to use it directly as a NUClear clock time point as it's misleading. Unless you know for sure that the time is measured in the same clock you'll have to do a time conversion somehow or just send it as a duration of some sort.